Feature spec : Authoring and deploying UDT (#69)
radius-project/design-notes: cli/2024-10-cli-udt.md

> # Topic: CLI commands for User-Defined Types

Author: ```suggestion
# CLI commands for User-Defined Types
```

The key scenarios for UDTs are outlined [here](/architecture/2024-06-resource-extensibility-feature-spec.md). We will decompose these scenarios specific to the CLI experience.

### Scenario 1: Deb defines and deploys the internal messaging service using the CLI
<!-- One or two sentence summary -->
Deb, a platform engineer who wants to define and deploy their internal messaging service Plaid as a resource type in Radius. 

Before we start designing the experience, we need to understand the high level terminologies/concept that a user would interact with in the CLI experience.

| Terminology | Description | Example |
|-------------|-------------|---------|
|Resource provider | Refers to a group of resource types under a common namespace. Resource providers are also entities that implement the group of resource types | Mycompany.Messaging or Applications.Core |
|Resource type | A service or technology that can be modeled as a resource | PostgreSQL or AWS S3 or Internal messaging service |
|Resource | An instance of a resource type | mydb or mybucket |

#### User Experience

> 1. Deb uses the rad CLI to scaffold a `yaml` template definition for the internal messaging service `Plaid`

Reviewer: Is this step really necessary? I'd prefer that authoring a resource type is so simple that it does not require a template. I'm imagining if I want to create a mycompany.com/database that I only need a single YAML file. If that's the case, I can just copy/paste from the sample in the docs.

If the scaffolding is indeed creating multiple files, I think we should challenge ourselves to simplify that further.

Author: Yes I think its helps user get a schema template with basic properties of the resource type from a rad command. Its convenience for the user over manually creating and copy pasting the template from docs. 

We do this already to scaffold an application(app.bicep) when you do `rad init`

Author: Ok. I will move this to commands to consider for future iterations

    Option 1:

    ```bash
    rad resource-provider init Mycompany.Messaging --schema-format yaml
    ```
        
    This command will create a folder with the name `Mycompany.Messaging`. A sample `yaml` file is created in the application folder. If we enable typespec, a sample `tsp` will be created and all the required dependencies will be created.

    Pros : Depicts the action of initializing the project with directory structure and dependencies. TypeSpec has similar `init` command to set up the project and would be consistent with that when we support typespec.

    Option 2:

    ```bash
    rad resource-type new Mycompany.Messaging --schema-format yaml
    ```
    
    Deb updates the schema definition with the required fields and properties for the messaging service Plaid. The authoring experience for the resource type will be covered in a separate document.

> 2. Deb registers the resource type with Radius

Reviewer: I'm a confused between resource types and resource providers. Back to my mycompany.com/database example, when I create the resource provider, what is actually being created? I think it's just a namespace. 

I'm guessing that you are saying that I create the resource provider and all resource types within the resource provider folder are created at once. If that's the case, I think that's the wrong direction. The user should be explicit about what resource types they are creating. I'd propose `rad resource-type create -f mydatabase.yaml`

Author: A `rad resource-provider init` would scaffold a structure dir like below

Mycompany.database
|-------> mydatabase.yaml

an organization can have have mutiple resource types 

Mycompany.database
|------->mydatabase.yaml
|------->resource2.yaml

when they create the resource-provider, all the resource types under the resource provider are registered to Radius. 

I also think that the user would want to add a resource type directly to a resource provider. Thats highlighted in option 2 of step 5

Author: I vote for having all resource types for a resource provider in the same file

    Option 1 

    ```bash
    rad resource-provider create Mycompany.Messaging
    ```
    Pros : Follows the CRUDL convention and is more consistent with the other commands in Radius

    Option 2

    ```bash

>     rad resource-provider register Mycompany.Messaging

Reviewer: I'd love for us to have consistent verbs. I would propose that `create` is the verb for creating new objects within Radius. `register` is for linking to existing things, like registering a cloud provider. It's unfortunate that `rad recipe register` is done that way since it's inconsistent with this definition.

Reviewer: That helps me understand the thinking on recipe register. Thanks.

    ```
    Pros : Consistent with Recipe template registration to environment
    Cons : Does not follow the CRUDL convention

3. Deb creates the resource instance for the messaging service Plaid

    ```bash
    rad resource create Mycompany.Messaging/Plaid -f myplaid.json
    ```
    The resource `myplaid` of type `Mycompany.Messaging/Plaid`is created and available for use.

4. Deb lists the resource providers and resource types available in Radius

    ```bash

>     rad resource-provider list

Reviewer: Is this necessary to implement? I doubt the number of resource types is going to be that big that the user can't just use `rad resource-type list`.

Author: There should be a way in CLI for them to query what resource types are added or available. Pagination could be something we support later on if the resource types list becomes huge.

    ```
    The list of resource providers in Radius are displayed.

    ```bash
    rad resource-type list --resource-provider Mycompany.Messaging
    ```
    The list of resource types under the resource provider `Mycompany.Messaging` are displayed.

> 5. Deb makes changes to the resource type definition of `Plaid` and adds to the resource provider

Reviewer: We probably need to talk about versioning here. Is a resource type mutable? If so, what do we do with existing resources already provisioned? Maybe it's immutable but has a version.

Reviewer 2: I'm in favor of making api versions mutable, but the schema of an api version immutable. This is a policy that ARM follows and I think it works decently. 

This allows non-breaking changes to a published API - things like doc improvements.

    Option 1:

    ```bash
    rad resource-provider create Mycompany.Messaging
    ```
    Pros : Same command can be used to update the resource provider with the new or updated resource type definitions

    Option 2 :

    ```bash

>     rad resource-type create plaid.yaml 

Reviewer: It doesn't seem like we have a consistent way of referring to an outside file in the CLI. Should it be -f? @? Or maybe a flag isn't necessary at all? I just think we should be consistent. We use -f earlier.

Author: yeah there should be a `-f` here.

>     rad resource-type create plaid.yaml 

Author: ```suggestion
    rad resource-type create -f plaid.yaml 
```

    ```
    Pro : No need to follow a rp directory structure. Resource type gets created/updated to the resource provider. 
    Con : Resource provider is a required input in the `plaid.yaml` schema definition

    Comment: It may be good and complete to support both these options.

6. Deb wants to know the details of resource provider

    ```bash
    rad resource-provider show Mycompany.Messaging
    ```
    The details of the resource provider `Mycompany.Messaging` are displayed.

> 7. Deb deletes a resource-type Plaid from the resource provider Mycompany.Messaging

Reviewer: Do we need a deprecation command to prevent new resources from being created? It seems like if we're going to have a version we would want to automatically deprecate the older version.

    ```bash
    rad resource-type delete Mycompany.Messaging/Plaid
    ```
    The resource type `Plaid` is deleted from the resource provider `Mycompany.Messaging`

8. Deb deletes the resource provider Mycompany.Messaging

    ```bash
    rad resource-provider delete Mycompany.Messaging
    ```

## Finalized set of commands 

rad resource-provider 

rad resource-type

> rad resource create

Reviewer 2: Should we separate `rad resource create` from this doc? It's not really a UDT feature. It's just functionality we're missing that we happen to **need** now.

    ```
    The resource type `Plaid` is deleted from the resource provider `Mycompany.Messaging`

8. Deb deletes the resource provider Mycompany.Messaging

    ```bash
    rad resource-provider delete Mycompany.Messaging
    Warning: `myapp` application has resources of type `Mycompany.Messaging/plaidResource` provisioned and running. Are you sure you want to delete the resource provider Mycompany.Messaging? (yes/no)
    ```

## Finalized set of commands

### rad resource-provider

#### 1. rad resource-provider create

Resource providers are entities that implement a group of resource types such as `Applications.Core`. Creating a resource provider involves creating new resource types that can be used in your applications. Inputs can be passed in as a file or as flags.

1. Create a resource provider from a file 
    
    ```bash
    rad resource-provider create -f Mycompany.Messaging.yaml
    Resource provider Mycompany.Messaging created successfully
    ```

2. Create a resource provider with flags

    ```bash

>     rad resource-provider create --name Mycompany.Messaging 

Reviewer 2: Did you write a scenario for this one? I'm just wondering what happens next 🤔 if I go down this path instead of `1.` (writing a manifest).

    Resource provider Mycompany.Messaging created successfully
    ```

**Flags**
- `-f`, `--from-file` : File path to the schema definition file
- `-n`, `--name` : Name of the resource provider

Question:

> - Should there be a flag that indicates the tenant to which the resource provider is created? Might be needed for multi

Reviewer 2: Can we leave this until we have planned work for multi-tenancy?

    ```

#### 4. rad resource-provider delete
Resource providers are entities that implement a group of resource types such as `Applications.Core`. Users can delete a resource provider.

    ```bash
    rad resource-provider delete Mycompany.Messaging
    ```
**Behaviors**
Option 1 : Radius can prompt the user with a warning message if there are resources provisioned and running under the resource provider. If the user confirms deletion, Radius cleans up everything, deletes resources, types and resource providers.

Option 2 : Radius can error out if there are resources provisioned and running under the resource provider. The user needs to delete the resources before deleting the resource provider.

### rad resource-type

#### 1. rad resource-type create
Resource types are the entities that implement resource types such as 'Applications.Core/containers'. Each resource type can define multiple API versions, and each API version defines a schema that resource instances conform to. Inputs can be passed in as a file or as flags.

1. Create a resource type from a file 
    
    ```bash
    rad resource-type create -f plaid.yaml
    ```

2. Create a resource type from flags

    ```bash

>     rad resource-type create --resource-provider Mycompany.Messaging --resource-type plaidResource --version=2024-10-09 --properties host=string port=string

Reviewer 2: I'm skeptical of this idea because I don't want to build two authoring experiences. It's going to be very hard to replicate the power of the written YAML manifest.

Is is **absolutely critical** to have?

Author: Agree that this is not a scalable solution. We could do interactive experiences in the future based on user feedback and adoption

The key scenarios for UDTs are outlined [here](/architecture/2024-06-resource-extensibility-feature-spec.md). We will decompose these scenarios specific to the CLI experience.

### Scenario 1: Deb defines and deploys the internal messaging service using the CLI
<!-- One or two sentence summary -->
Deb, a platform engineer who wants to define and deploy their internal messaging service Plaid as a resource type in Radius. 

Before we start designing the experience, we need to understand the high level terminologies/concept that a user would interact with in the CLI experience.

| Terminology | Description | Example | 
|-------------|-------------|---------|
|Resource provider | Refers to a group of resource types under a common namespace. Resource providers are also entities that implement the group of resource types | Mycompany.Messaging or Applications.Core |
|Resource type | A service or technology that can be modeled as a resource | PostgreSQL or AWS S3 or Internal messaging service |
|Resource | An instance of a resource type | mydb or mybucket |

> #### User Experience

Author: Refine the user exp to include the commands that a user would use 
1. rad resource-type create
2. call out support for both yaml and json

Below is an illustrative example of how Deb would define the internal messaging service Plaid as a resource type in Radius. 

1. Deb creates the resource provider and the resource type with Radius

    ```bash

>     rad resource-provider create -f Mycompany.Messaging.yaml

Author: Updated this to be consistent across commands

    ```
    `Mycompany.Messaging.yaml` is the schema definition file that contains the resource type definition for the messaging service Plaid. See the feature spec that covers the [authoring experience for the schema definition](https://github.com/radius-project/design-notes/blob/67dd03056d22bcae2ee9743ffb308c8ac7f012bd/architecture/2024-10-user-defined-types-authoring-experience-feature-spec.md)

2. Deb creates the resource instance for the messaging service Plaid

    ```bash

>     rad resource create Mycompany.Messaging/Plaid myplaid -f myplaid.json

Reviewer 2: One thine we could consider - is whether we want to encode the type and resource name in the file vs at the command line. 

I don't have a strong preference right now. I think it's a worthwhile discussion.

Author: As decided in the design meeting, updated this to have type and resource name in the file.

>     rad resource create Mycompany.Messaging/Plaid myplaid -f myplaid.json

Reviewer 3: Does the resource type `Plaid` get created as a part of the previous resource provider create command?

Author: yes

Reviewer 4: I'm a bit confused on this as well. We create the resource type as part of the previous command but then the resource instance here?

Author: yes `rad resource-type create` creates the resource type definition and `rad resource create` creates the resource instance.

    ```

>     The resource `myplaid` of type `Mycompany.Messaging/plaidResource`is created and available for use.

Reviewer 3: Should this be Mycompany.Messaging/Plaid as indicated in the command arg in the line above?
```suggestion
    The resource `myplaid` of type `Mycompany.Messaging/Plaid`is created and available for use.
```

Author: thanks for catching that it shd either be Mycompany.Messaging/plaidResource or Mycompany.Messaging/plaid

Author: Updated this to follow Mycompany.Messaging/plaidService

4. Deb lists the resource providers and resource types available in Radius

    ```bash
    rad resource-provider list
    ```
    The list of resource providers are displayed.

    ```bash
    rad resource-type list --resource-provider Mycompany.Messaging
    ```
    The list of resource types under the resource provider `Mycompany.Messaging` are displayed.

> 5. Deb makes changes to the resource type definition of `plaidResource` and adds to the resource provider

Reviewer 4: Just to clarify, there won't be a specific `update` command. So in the update scenario, do we expect the existing definitions to get deleted and then overwritten?

Author: Yes there won't be specific update command. It depends on what kind of change (breaking vs non-breaking), the user is making which is covered here https://github.com/radius-project/design-notes/pull/74

    ```bash

>     rad resource-provider create -f Mycompany.Messaging.yaml

Reviewer 5: three questions: 
1. Does Deb need to specify the name of the resource type (i.e. `plaidResource`) as a part of this command?
2. Are there any thoughts about overloading `create` to perform both creation and update operations vs. adding another command for `update`?
3. Should this be a `rad resource-type create` command instead?

Author: 1. Yes updated this to use `rad resource-type` as its more intuitive and we expect users to initially interact via that commans
2. Create is both create and update. This is consistent with other commands in Radius. We do not have separate commands for update.
3. Yes

>     ```

Reviewer 3: So updates will be performed via create as an override?

Author: yes consistent with other commands in Radius

> 6. Deb wants to know the details of resource provider

Reviewer 5: Should there also be a `show` command for a specific resource type?

Reviewer 5: pls ignore, i see this is addressed in a subsequent section.

    ```bash

>     rad resource-provider show Mycompany.Messaging

Reviewer 3: Couple of questions -
1. What's the motivation to make RP name a positional arg here but a flag in the list command?
2. What does the output here include in addition to what's displayed in the list command output?

    ```

>     The details of the resource provider `Mycompany.Messaging` are displayed.

Author: below is what we want to show to begin with 
RESOURCE-PROVIDER   TYPE                                                          STATE
Mycompany.Messaging   System.Resources/resourceProviders  Succeeded

7. Deb deletes a resource-type Plaid from the resource provider Mycompany.Messaging

    ```bash
    rad resource-type delete Mycompany.Messaging/plaidResource

>     Warning: `myapp` application has resources of type `Mycompany.Messaging/plaidResource` provisioned and running. Are you sure you want to delete the resource type plaidResource from the resource provider Mycompany.Messaging? (yes/no)

Reviewer 6: does this delete the provisioned resources as well or no?

Reviewer 6: I see you addressed this below

Author: updated to include the final decided behavior

    ```
    The resource type `Plaid` is deleted from the resource provider `Mycompany.Messaging`

8. Deb deletes the resource provider Mycompany.Messaging

    ```bash

>     rad resource-provider delete Mycompany.Messaging

Author: We can delete the built in resource providers, we need a way to reload the built-in types from a url

    Warning: `myapp` application has resources of type `Mycompany.Messaging/plaidResource` provisioned and running. Are you sure you want to delete the resource provider Mycompany.Messaging? (yes/no)
    ```

> ## Finalized set of commands

Reviewer 2: I think you might have some unbalanced backticks, this is showing up in grey.

### rad resource-provider

#### 1. rad resource-provider create

Resource providers are entities that implement a group of resource types such as `Applications.Core`. Creating a resource provider involves creating new resource types that can be used in your applications. Inputs can be passed in as a file or as flags.

Create a resource provider from a file 
    
    ```bash

>     rad resource-provider create -f Mycompany.Messaging.yaml

Reviewer 2: Would it make sense to support URLs here as well? eg: a link to a GH repo. Perhaps **our** github repo 😁

    Resource provider Mycompany.Messaging created successfully
    ```
**Flags**
- `-f`, `--fom-file` : The file that contains the resource provider definition

#### 2. rad resource-provider list
Resource providers are entities that implement a group of resource types such as `Applications.Core`. Users can list all the resource providers available in your tenant.

    ```bash
    rad resource-provider list 
    RESOURCE-PROVIDER   TYPE                                STATE
    Mycompany.Messaging System.Resources/resourceProviders  Succeeded
    ```

#### 3. rad resource-provider show
Resource providers are entities that implement a group of resource types such as `Applications.Core`. User can display the details of the resource providers.

    ```bash
    rad resource-provider show Mycompany.Messaging

>     RESOURCE-PROVIDER   TYPE                                STATE

Reviewer 2: Today, the column heading for the `name` is always `RESOURCE` regardless of the resource type. We could change this in the future, but `RESOURCE` would be consistent with our other commands.

Author: check this across other commands - app, env, workspace, resources

    Mycompany.Messaging System.Resources/resourceProviders  Succeeded
    ```

#### 4. rad resource-provider delete
Resource providers are entities that implement a group of resource types such as `Applications.Core`. Users can delete a resource provider.

    ```bash
    rad resource-provider delete Mycompany.Messaging
    ```
**Behaviors**
Option 1 : Radius can prompt the user with a warning message if there are resources provisioned and running under the resource provider. If the user confirms deletion, Radius cleans up everything, deletes resources, types and resource providers.

Option 2 : Radius can error out if there are resources provisioned and running under the resource provider. The user needs to delete the resources before deleting the resource provider.

> 

Reviewer 2: @Reshrahim and I have had a few discussions about this. We should nail down both:

- The behavior of the API: The API has to enforce invariants.
- The behavior of the CLI: The CLI can provide convenience for the user.

These don't have to be 100% the same, but they have to be largely in agreement. 

Any behaviors that we want to consistent for all clients, should probably be enforced at the API layer.

Author: For deleting a resource type or resource provider, we decided to do the following :
    1. API - Error out of if there are resources provisioned with resource type 
    2. CLI - Prompt the user with a warning message that there are resources provisioned with the resource type. If the user confirms deletion, API behavior is exercised.

### rad resource-type

#### 1. rad resource-type create
Resource types are the entities that implement resource types such as 'Applications.Core/containers'. Each resource type can define multiple API versions, and each API version defines a schema that resource instances conform to. Inputs can be passed in as a file or as flags.

Create a resource type from a file 
    
```bash
rad resource-type create Mycompany.Messaging/plaidResource -f plaid.yaml

> ```

Reviewer 2: Do we need both `rad resource-provider create` and `rad resource-type create`? what would happen if we just built one of these?

Also, is this the same file as was used in `rad resource-provider create`?

As much as possible, let's create shared investments.

Author: yes same behavior across `rad resource-provider create` and `rad resource-type create`. We expect `rad resource-type create` to be the command a user interacts with and will be featured in the docs

```bash
rad resource-type list --resource-provider Mycompany.Messaging
RESOURCE-TYPE   TYPE   STATE
plaidResource   Mycompany.Messaging Succeeded
resourceType2   Mycompany.Messaging Succeeded
```

#### 3. rad resource-type show
Resource types are the entities that implement resource types such as 'Applications.Core/containers'. Users can display the details of the resource types.

```bash
rad resource-type show Mycompany.Messaging/plaidResource
RESOURCE-TYPE   TYPE   STATE
plaidResource   Mycompany.Messaging Succeeded
```

#### 4. rad resource-type delete
Resource types are the entities that implement resource types such as 'Applications.Core/containers'. Users can delete a resource type.

```bash
rad resource-type delete Mycompany.Messaging/plaidResource
```

**Behaviors**:

Option 1 : Radius can prompt the user with a warning message if there are resources provisioned and running under the resource type. If the user confirms deletion, Radius cleans up everything, deletes resources, types and resource providers.

> Option 2 : Radius can error out if there are resources provisioned and running under the resource type. The user needs to delete the resources before deleting the resource type.

Reviewer 2: Note that there's inherently a data-race at work here if we implement this behavior in the CLI.

If other users are creating resources at the same thing, then this command might fail. I think that's OK, it's an edge case and we prevented you from data loss.

Deb, a platform engineer who wants to define and deploy their internal messaging service Plaid as a resource type in Radius. 

Before we start designing the experience, we need to understand the high level terminologies/concept that a user would interact with in the CLI experience.

| Terminology | Description | Example | 
|-------------|-------------|---------|
|Resource type | A service or technology that can be modeled as a resource | PostgreSQL or AWS S3 or Internal messaging service |
|Resource provider | Refers to a group of resource types under a common namespace. Resource providers are also entities that implement the group of resource types | Mycompany.Messaging or Applications.Core |
|Resource | An instance of a resource type | mydb or mybucket |

#### User Experience
Below is an illustrative example of how Deb would define and deploy the internal messaging service Plaid as a resource in Radius. 

1. Deb defines the schema for plaid messaging service in a `plaid.yaml` file. Note this schema definition is illustrative and may not be the final schema definition. This will evolve in [authoring experience feature spec](https://github.com/radius-project/design-notes/pull/74)

>     ```yaml

Author: > Two main questions I have though:
> 
> 1. Is it necessary for the user to specify application and environment properties? This seems like unnecessary boilerplate stuff. Radius knows it's a resource type so it should know to add those properties.

It indicate the scope of the resources whether is application or environment scoped. This is a bigger design change to make across all resource types in Radius not just UDT. We have tracking issue here - https://github.com/radius-project/radius/issues/6398

> 2. Similar question on capabilities. Is this just boilerplate config? What's the user functionality for these things? I don't think we should be asking a user in a spec unnecessary for information that is not used.

Capabilities define the behaviors of the resource. Some of the behaviors are 
1. how do you want the resource type to be provisioned ? via Recipes or via some other provisioning logic
2. How should someone connect with me ? What is my connection information etc

@rynowak  please add anything I might have missed.

                                type: 'string'
                                description: 'Application name'
                            environment:
                                type: 'string'
                                description: 'Environment name'
                            queueName:
                                type: 'string'
                                description: 'Name of the queue'
                            host:
                                type: 'string'
                                description: 'Hostname of the messaging service'        
                            port:
                                type: 'string'
                                description: 'Port'
                            connectionString:
                                type: 'string'
                                description: 'Connection string to the messaging service' 
                        required: ['queueName', 'host', 'port', 'connectionString']
                    capabilities: ["Recipe"]           
    ....
    ```

1. Deb creates the resource type for the messaging service Plaid in Radius

    ```bash
    rad resource-type create -f plaid.yaml
    ```

> 2. Deb creates the resource instance for the messaging service Plaid

Author: How will Deb know his resource definition works for his developer?.This is similar to a tutorial to see the working of UDT end-end

    ```json
    {
        "name": "myplaid",
        "type": "Mycompany.Messaging/plaidService",
        "properties": {
            "application": "myapp"
            "environment": "dev"
            "queueName": "myqueue"
            "host": "plaidxyz.mycompany.com"
            "port": "5672"
    ...
    }
    ```

    ```bash

>     rad resource create -f myplaid.json

Author: For `rad deploy` or `rad run` to work, we need the Bicep tooling support. This helps to create resource instances for testing until then.

    ```
    The resource `myplaid` of type `Mycompany.Messaging/plaidService`is created and available for use.

4. Deb lists the resource types available in Radius

    ```bash
    rad resource-type list
    ```
    The list of resource types under the resource provider `Mycompany.Messaging` are displayed.

5. Deb makes changes to the resource type definition of `Mycompany.Messaging/plaidService`

    ```bash
    rad resource-type create -f plaid.yaml
    ```
    Note that the experience for handling breaking changes and non-breaking changes are detailed in the [authoring experience feature spec]((https://github.com/radius-project/design-notes/pull/74)

6. Deb wants to know if the resource-type is created successfully

    ```bash
    rad resource-type show Mycompany.Messaging/plaidService
    ```
    The details of the resource type `Mycompany.Messaging` are displayed.

7. Deb deletes a resource-type Plaid from the resource provider Mycompany.Messaging

    ```bash

>     rad resource-type delete Mycompany.Messaging/plaidService

Reviewer: ```suggestion
    rad resource-type delete Mycompany.Messaging/plaidService@2024-10-01
```

>     Warning: `myapp` application has resources of type `Mycompany.Messaging/plaidResource` provisioned and running. Are you sure you want to delete the resource type plaidResource from the resource provider Mycompany.Messaging? (yes/no)

Reviewer: This warning need to be updated. See comment below.

    ```

>     The resource-type deletion errors out as there are resources provisioned and running under the resource type. The resource myplaid should be deleted before deleting the resource type. 

Reviewer: I disagree with this behavior. Here is my use case:
1. Platform engineer creates Mycompany.Messaging/plaidService@2024-10-01
2. Developer create resources using the 2024-10-01 version
3. Platform engineer creates Mycompany.Messaging/plaidService@2024-11-15
4. Platform engineer wants to restrict new resources being created with version 2024-10-01 so deletes resource type Mycompany.Messaging/plaidService@2024-10-01
5. Existing resources remain deployed

## Finalized set of commands

### rad resource-type

#### 1. rad resource-type create
Resource types are the entities that implement resource types such as 'Applications.Core/containers'. Each resource type can define multiple API versions, and each API version defines a schema that resource instances conform to. Inputs can be passed in as a file or as flags.

Create a resource type from a file 
    
```bash

> rad resource-type create Mycompany.Messaging/plaidResource -f plaid.yaml

Reviewer: ```suggestion
rad resource-type create -f plaid.yaml
```

Reviewer: The resource name is in the yaml file

Reviewer: plaidResource isn't the resource type name.

```

**Flags**

> - `-f`, `--fom-file` : The file that contains the resource type definition

Reviewer: Not sure if this is a typo, but --filename seems more commonly used.

#### 2. rad resource-type list
Resource types are the entities that implement resource types such as 'Applications.Core/containers'. Users can list all the resource types available in the resource provider.

```bash

> rad resource-type list --resource-provider Mycompany.Messaging

Reviewer: ```suggestion
rad resource-type list --namespace Mycompany.Messaging
```

NAME            TYPE                STATE

> plaidResource   Mycompany.Messaging Succeeded

Reviewer: ```suggestion
NAME                                            TYPE             STATE
Mycompany.Messaging/plaidServices@2024-10-01    Resource Type    Available
Mycompany.Messaging/plaidServices@2024-11-15    Resource Type    Available
```

resourceType2   Mycompany.Messaging Succeeded
```

#### 3. rad resource-type show

> Resource types are the entities that implement resource types such as 'Applications.Core/containers'. Users can display the details of the resource types.

Reviewer: ```suggestion
Display details of the resource type including versions and schemas. Resource type parameter can optionally include version.
```

```bash

> rad resource-type show Mycompany.Messaging/plaidResource

Reviewer: ```suggestion
rad resource-type show Mycompany.Messaging/plaidServices@2024-10-01
```

NAME            TYPE                STATE

> plaidResource   Mycompany.Messaging Succeeded

Reviewer: ```suggestion
Mycompany.Messaging/plaidServices
Version: 2024-10-01
queueName        string Required Name of the queue
host             string Required Hostname of the messaging service
port             string Required port
connectionString string Required Connection string to the messaging service

Version: 2024-11-15
queueName        string Required Name of the queue
host             string Required Hostname of the messaging service
port             string Required port
connectionString string Required Connection string to the messaging service
newParameter     string          New optional parameter
```

```

> #### 4. rad resource-type delete

Reviewer: See comments above about this behavior. I don't think we're taking into account changes to the schema and versions.

Resource types are the entities that implement resource types such as 'Applications.Core/containers'. Users can delete a resource type.

```bash
rad resource-type delete Mycompany.Messaging/plaidResource
```
We decided to prompt the user with a warning message if there are resources provisioned and running under the resource type. The resource-type deletion errors out as there are resources provisioned and running under the resource type. The resource myplaid should be deleted before deleting the resource type. 

> ### rad resource create

Reviewer: Why does this exist? What's the use case?

Resources are instances of resource types such as 'Applications.Core/containers'. Inputs can be passed in as a file or as flags.

Create a resource from a file 
    
```bash
rad resource create -f myplaid.json
```

**Flags**
- `-f`, `--fom-file` : The file that contains the resource properties

> ### rad resource-provider

Reviewer: Why does this exist? What's the use case? I thought we agreed to not surface this in the CLI UX?
